qiita_db/sql_connection.py

```diff
@@ -22,57 +22,144 @@
-    "phone) VALUES (%s, %s, %s, %s) RETURNING email, password",
-    ['insert@foo.bar', 'Toy', 'pass', '111-111-11112']) # doctest: +SKIP
-conn_handler.add_to_queue(
-    "example_queue", "UPDATE qiita.qiita_user SET user_level_id = 1, "
-    "phone = '222-222-2221' WHERE email = %s AND password = %s",
-    ['{0}', '{1}']) # doctest: +SKIP
-conn_handler.execute_queue("example_queue") # doctest: +SKIP
-conn_handler.execute_fetchall(
-    "SELECT * from qiita.qiita_user WHERE email = %s", ['insert@foo.bar'])
-[['insert@foo.bar', 1, 'pass', 'Toy', None, None, '222-222-2221', None, None,
-  None]] # doctest: +SKIP
+
+>>> from qiita_db.environment_manager import drop_and_rebuild_tst_database
+>>> drop_and_rebuild_tst_database()  # doctest: +SKIP
+
+* Querying
+
+In order to perform any query in the database you first need to import the
+`TRN` variable available in this module:
+
+>>> from qiita_db.sql_connection import TRN
+
+The `TRN` variable is an instance of the `Transaction` object. All queries
+should be executed through this object to ensure atomicity. To perform any
+query you first need to add the query to the transaction and then execute the
+transaction:
+
+>>> with TRN:
+...     TRN.add("SELECT 42")
+...     res = TRN.execute()
+>>> res
+[[[42]]]
+
+The `execute` function returns the values of all the queries in the transaction
+object. This requires three layers of nesting: (1) the query (2) the result
+rows in a given query and (3) the result columns in a given row.
+
+* Data retrieval
+
+Multiple auxiliary functions exist for easy SQL result retrieval:
+
+Getting the first value of the last SQL query
+>>> with TRN:
+...     TRN.add("SELECT 42")
+...     res = TRN.execute_fetchlast()
+>>> res
+42
+
+Getting the results of the last SQL query
+>>> with TRN:
+...     TRN.add("SELECT 42")
+...     res = TRN.execute_fetchindex()
+>>> res
+[[42]]
+
+Getting the results of the specified SQL query
+>>> with TRN:
+...     TRN.add("SELECT 42")
+...     TRN.add("SELECT 43")
+...     # The index 0 corresponds to the first query in the transaction
+...     res = TRN.execute_fetchindex(0)
+>>> res
+[[42]]
+
+Getting the results of the last SQL query flattened
+>>> with TRN:
+...     TRN.add("SELECT 42, 43, 44")
+...     res = TRN.execute_fetchflatten()
+>>> res
+[42, 43, 44]
+
+Getting the results of the specified SQL query flattened
+>>> with TRN:
+...     TRN.add("SELECT 42, 43, 44")
+...     TRN.add("SELECT 42")
+...     res = TRN.execute_fetchflatten(0)
+>>> res
+[42, 43, 44]
+
+* Transactions
+
+Transaction blocks are created through the same `TRN` variable exported in this
+module. You can add as many SQL commands as you want and execute all of them at
+once, and it will return the results of all the SQL commands. `TRN` should be
+used as a context so manager, and it autocommits the transaction once the last
+context is exited, as long as no error was generated inside the context, in
+which case a rollback is executed.
+
```

Comment:
Yes, by definition of transaction, any command added to the current transaction is undone. Note that everything that has been already committed it will not be included in the rollback. Ex:

```python
with TRN:
    # (A) do stuff

with TRN:
    # (B) do stuff
    raise ValueError
```
In the above example, the stuff in (A) is committed, but the stuff in (B) is rollbacked. On the other hand:
```python
with TRN:
    with TRN:
        # (A) do stuff

    with TRN:
        # (B) do stuff
        raise ValueError
```
neither (A) or (B) are committed, so both of them will be rollbacked.